### PR TITLE
Add mir_window_state_attached

### PR DESCRIPTION
**Pull request by wmww**
Add the `mir_surface_type_attached` window state, and two new properties to windows: `attached_edges` and `exclusive_rect`. These will be important for implementing layer surfaces. I still need to add tests, but I want to throw this up now so I can start getting feedback on it.

### CODE REVIEW

**Comment by wmww** [comment]
Now it's tested and working

File: src/miral/basic_window_manager.cpp [hunk 1]
```diff
@@ -1127,6 +1137,54 @@ void miral::BasicWindowManager::place_and_size(WindowInfo& root, Point const& ne
     move_tree(root, new_pos - root.window().top_left());
 }
 
+void miral::BasicWindowManager::place_attached(WindowInfo& root)
+{
+    // We can't place the attached window if there aren't any outputs
+    if (root.type() != mir_window_type_attached || outputs.size() == 0)
```
Comment (RAOF) [inline src/miral/basic_window_manager.cpp]: Likewise, `root.type() != mir_window_type_attached` here is a precondition failure.

File: src/miral/window_info.cpp [hunk 2]
```diff
@@ -174,10 +180,12 @@ bool miral::WindowInfo::can_be_active() const
     case mir_window_type_gloss:
     case mir_window_type_tip:          /**< AKA "tooltip"                       */
     case mir_window_type_decoration:
-    default:
+    case mir_window_type_attached:     /**< AKA "panel"/"notification"/"etc"    */
```
Comment (RAOF) [inline src/miral/window_info.cpp]: Can panel/notification windows *really* not have input focus? That seems odd!

**Comment by AlanGriffiths** [comment]
I've not read the code, but "attached" sounds more like a window state than a window type.

I can imagine "attached" being used for "normal", "gloss" or "inputmethod" windows types.

I can't imagine "attached" taking "fullscreen", "vertmaximized" or "minimized" window states.

**Comment by wmww** [comment]
Switched attached from a window type to a window state as per my discussion this morning with Alan. How's it looks now? If there are any concerstatens from before I've forgot to address, just reraise them.

**Comment by wmww** [comment]
Added another test and fixed the unused function warning in Clang

File: src/miral/window_info.cpp [hunk 3]
```diff
@@ -114,11 +116,15 @@ miral::WindowInfo::Self::Self(Window window, WindowSpecification const& params)
     min_aspect(optional_value_or_default(params.min_aspect(), default_min_aspect_ratio)),
     max_aspect(optional_value_or_default(params.max_aspect(), default_max_aspect_ratio)),
     shell_chrome(optional_value_or_default(params.shell_chrome(), mir_shell_chrome_normal)),
-    depth_layer(optional_value_or_default(params.depth_layer(), mir_depth_layer_application))
+    depth_layer(optional_value_or_default(params.depth_layer(), mir_depth_layer_application)),
+    attached_edges(optional_value_or_default(params.attached_edges(), mir_placement_gravity_center))
```
Comment (AlanGriffiths) [inline src/miral/window_info.cpp]: It isn't obvious why this should default to "center". Are we using that as a synonym for "none"?

Comment (wmww) [reply src/miral/window_info.cpp]: Attaching to the center of the output seems like as good a default as any. I don't think it's terribly important what the default is, as the value will generally be set at the same time as the state is changed to attached.

**Comment by wmww** [comment]
No changes, just fixed the conflict

File: src/miral/window_info.cpp (hunk 3 above)
Comment (RAOF) [reply src/miral/window_info.cpp]: Is it *mandatory* that the attachment should be set when switching to attached state? That sounds reasonable?

Comment (wmww) [reply src/miral/window_info.cpp]: No. If it was then that would imply it's mandatory to set the edges every time you switch into an attached state, even if you had already set the edges the previous time. Also, that would also imply it's impossible to change the attached edges when the window is not in an attached state, which in some cases may be convenient. I don't see the potential for problems with just having them be independent properties.

Comment (AlanGriffiths) [reply src/miral/window_info.cpp]: As `attached_edges` is always set to _something_ then not requiring it to be set when switching to "attached" is consistent. If that's what we elect to do then any arbitrary default value works and "center" is the canonical "middle of the road".

File: src/miral/symbols.map [hunk 4]
```diff
@@ -438,15 +438,19 @@ global:
     miral::WaylandExtensions::supported[*;
     miral::WaylandExtensions::zwlr_layer_shell_v1*;
     miral::WaylandExtensions::zxdg_output_manager_v1*;
+    miral::WindowInfo::attached_edges*;
```
Comment (AlanGriffiths) [inline src/miral/symbols.map]: Unless we plan on cherry-picking into 1.3.0 (I think that unnecessary) these need moving to 2.7

**Comment by wmww** [comment]
Symbols should be good to go now

**Comment by wmww** [comment]
bors r=AlanGriffiths

**Pull request merged**